File: Crtm05/GeographicCoordinate.cs
using System;

namespace Crtm05
{
    public class GeographicCoordinate
    {
        public double Coordinate
        {
            get
            {
                return coordinate;
            }
            set
            {
                coordinate = value;
                UpdateDegreesMinutesSeconds();
            }
        }

        public int Degrees
        {
            get
            {
                return degrees;
            }
            set
            {
                degrees = value;
                UpdateCoordinate();
            }
        }

        public int Minutes
        {
            get
            {
                return minutes;
            }
            set
            {
                minutes = value;
                UpdateCoordinate();
            }
        }

        public double Seconds
        {
            get
            {
                return seconds;
            }
            set
            {
                seconds = value;
                UpdateCoordinate();
            }
        }

        double coordinate;
        int degrees;
        int minutes;
        double seconds;

        public GeographicCoordinate(double coordinate)
        {
            Coordinate = coordinate;
        }

        public GeographicCoordinate(int degrees, int minutes, double seconds)
        {
            this.degrees = degrees;
            this.minutes = minutes;
            this.seconds = seconds;
            UpdateCoordinate();
        }

        void UpdateDegreesMinutesSeconds()
        {
            degrees = (int)Math.Floor(coordinate);
            minutes = (int)Math.Floor((coordinate - degrees) * 60);
            seconds = (coordinate - degrees - ((double)minutes / 60)) * 3600;
        }

        void UpdateCoordinate()
        {
            coordinate = degrees + ((double)minutes / 60) + (seconds / 3600);
        }
    }
}

File: typescript/Crtm05/Geographic/geographic-coordinate.ts
export class GeographicCoordinate {
  private coordinate: number
  private degrees: number
  private minutes: number
  private seconds: number

  constructor (coordinate?: number, degrees?: number, minutes?: number, seconds?: number) {
    if (coordinate) {
      this.coordinate = coordinate
      this.updateDegreesMinutesSeconds()
    } else {
      this.degrees = degrees
      this.minutes = minutes
      this.seconds = seconds
      this.updateCoordinate()
    }
  }

  getCoordinate (): number {
    return this.coordinate
  }

  setCoordinate (coordinate: number) {
    this.coordinate = coordinate
    this.updateDegreesMinutesSeconds()
  }

  getDegrees (): number {
    return this.degrees
  }

  setDegrees (degrees: number) {
    this.degrees = degrees
    this.updateCoordinate()
  }

  getMinutes (): number {
    return this.minutes
  }

  setMinutes (minutes: number) {
    this.minutes = minutes
    this.updateCoordinate()
  }

  getSeconds (): number {
    return this.seconds
  }

  setSeconds (seconds: number) {
    this.seconds = seconds
    this.updateCoordinate()
  }

  private updateDegreesMinutesSeconds () {
    this.degrees = Math.floor(this.coordinate)
    this.minutes = Math.floor((this.coordinate - this.degrees) * 60)
    this.seconds = (this.coordinate - this.degrees - this.minutes / 60) * 3600
  }

  private updateCoordinate () {
    this.coordinate = this.degrees + this.minutes / 60 + this.seconds / 3600
  }
}
